packages/doenetml-worker-rust/lib-doenetml-core/src/utils/test_utils.rs

```diff
@@ -0,0 +1,57 @@
+use crate::{
+    components::prelude::{Dependency, StateVarMutableView, StateVarView, StateVarViewEnum},
+    dependency::DependencySource,
+    state::essential_state::EssentialDataOrigin,
+};
+
+/// Create a mutual view of a state variable along with a read only view
+pub fn create_view_pair<T: Default + Clone>(
+    value: T,
+    came_from_default: bool,
+) -> (StateVarMutableView<T>, StateVarView<T>) {
+    let mutable_view = StateVarMutableView::new_with_value(value, came_from_default);
+    let view = mutable_view.create_new_read_only_view();
+    (mutable_view, view)
+}
+
+pub fn create_essential_dependency<T: Default + Clone>() -> (Dependency, StateVarMutableView<T>)
+where
+    // make sure T is one of the types that we can convert from StateVarView<T> into StateVarViewEnum
+    StateVarViewEnum: From<StateVarView<T>>,
+{
+    let (essential, essential_view) = create_view_pair(T::default(), true);
+
+    let dependency_source = DependencySource::Essential {
+        component_idx: 0,
+        origin: EssentialDataOrigin::StateVar(0),
+    };
+
+    let dependency = Dependency {
+        source: dependency_source,
+        value: essential_view.into(),
+    };
+
+    (dependency, essential)
+}
+
+pub fn create_state_var_dependency<T: Default + Clone>(
+    initial_value: T,
+) -> (Dependency, StateVarMutableView<T>)
+where
+    // make sure T is one of the types that we can convert from StateVarView<T> into StateVarViewEnum
+    StateVarViewEnum: From<StateVarView<T>>,
+{
+    let (state_var, state_var_view) = create_view_pair(initial_value, false);
+
+    let dependency_source = DependencySource::StateVar {
+        component_idx: 0,
```

Comment:
Literal indices scare me.

Reply:
If you want a test at this level without `core`, we need to specify values for these literals, as the interface between `core` and state variables is defined by these indices. I did end up removing the literal indices in the code that component developers write, but core's representation of components and state variables is in terms of such indices. Integration tests involving code from `core` could presumably avoid such indices.